### PR TITLE
[SPARK-28699][Core] Cache an indeterminate RDD could lead to incorrect result while stage rerun

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

It's another case for the indeterminate stage/RDD rerun while stage rerun happened. In the CachedRDDBuilder, we miss tracking the `isOrderSensitive` characteristic to the newly created MapPartitionsRDD.
This patch just a safeguard, if we need the support for stage rerunning, it should be done after #24892.

## How was this patch tested?

Integrated test with an exception, instead of the wrong answer.
```
import scala.sys.process._
import org.apache.spark.TaskContext

val res = spark.range(0, 10000 * 10000, 1).map{ x => (x % 1000, x)}
// kill an executor in the stage that performs repartition(239)
val df = res.repartition(113).cache.repartition(239).map { x =>
  if (TaskContext.get.attemptNumber == 0 && TaskContext.get.partitionId < 1 && TaskContext.get.stageAttemptNumber == 0) {
    throw new Exception("pkill -f -n java".!!)
  }
  x
}
val r2 = df.distinct.count()
```

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 1]
```diff
@@ -2710,7 +2710,7 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     assert(countSubmittedMapStageAttempts() === 2)
   }
 
-  test("SPARK-23207: retry all the succeeding stages when the map stage is indeterminate") {
+  ignore("SPARK-23207: retry all the succeeding stages when the map stage is indeterminate") {
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: Ignore this for the behavior change, as the approach now, we need to abort the stage of the current mapStage.
As we will finally support stage rerun, I suggest to skip this behavior, we can directly support the cache scenario after SPARK-25341 merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 2]
```diff
@@ -131,7 +131,7 @@ case class CachedRDDBuilder(
 
         def hasNext: Boolean = rowIterator.hasNext
       }
-    }.persist(storageLevel)
+    }, isOrderSensitive = true).persist(storageLevel)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: I don't quite get it. The `isOrderSensitive` flag is used to describe the map function. Why the map function is order sensitive?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 3]
```diff
@@ -86,7 +86,7 @@ case class CachedRDDBuilder(
 
   private def buildBuffers(): RDD[CachedBatch] = {
     val output = cachedPlan.output
-    val cached = cachedPlan.execute().mapPartitionsInternal { rowIterator =>
+    val cached = cachedPlan.execute().mapPartitionsWithIndexInternal({ (_, rowIterator) =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: I think we only run cached plan once. I think it should be determinate? If so, this map should not be order sensitive effectively?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
@xuanyuanking Can we have a test case to prevent a future regression on this?

**Comment by xuanyuanking** [comment]
Sorry for the late. Thanks for the comment and review! After taking a further investigation, I found the root cause for this issue is about radix sort in ShuffleExchangeExec, please review it in #25491.
The bug fix of DAGScheduler is still needed, I'll give another follow-up PR and add UT for it.

**Pull request closed without merging**